test/e2e_node/image_pull_test.go

```diff
@@ -0,0 +1,286 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2enode
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	kubeletevents "k8s.io/kubernetes/pkg/kubelet/events"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	"k8s.io/kubernetes/test/e2e/nodefeature"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+	"k8s.io/utils/ptr"
+)
+
+var _ = SIGDescribe("Pull Image", framework.WithSerial(), nodefeature.MaxParallelImagePull, func() {
+
+	f := framework.NewDefaultFramework("parallel-pull-image-test")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	httpdImage := imageutils.GetE2EImage(imageutils.Httpd)
+	httpdNewImage := imageutils.GetE2EImage(imageutils.HttpdNew)
+	var testpod, testpod2 *v1.Pod
+
+	ginkgo.Context("parallel image pull with MaxParallelImagePulls=5", func() {
+
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.SerializeImagePulls = false
+			initialConfig.MaxParallelImagePulls = ptr.To[int32](5)
+		})
+
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			testpod = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "httpd",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:            "httpd",
+						Image:           httpdImage,
+						ImagePullPolicy: v1.PullAlways,
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			}
+			testpod2 = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "httpd2",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:            "httpd-new",
+						Image:           httpdNewImage,
+						ImagePullPolicy: v1.PullAlways,
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			}
+
+			ginkgo.By("cleanup images")
+			_ = RemoveImage(testpod.Spec.Containers[0].Image)
+			_ = RemoveImage(testpod2.Spec.Containers[0].Image)
```

Comment:
Would it make sense to remove httpdImage and httpdNewImage explicitly or remove all container images by looping over all pods and containers and removing all found images?

Reply:
> Would it make sense to remove httpdImage and httpdNewImage explicitly or remove all container images by looping over all pods and containers and removing all found images?

I am not sure why we should touch other images than httpd images.

Do you mean that other images may share same base image layers or other special scenarios?